### ISSUE TITLE
[Docs] Introduce reference documentation for cython

### ISSUE THREAD

**Issue by matusvalo** [cython/cython#5849]
### Is your feature request related to a problem? Please describe.

Currently Cython has 2 types of documentation:
* tutorials
* userguides
* faq

There are several features not documented by Cython:
* supported data types (#5848)
* tempita templating system
* grammar definition
* ...

I think that current documentation has no place for mentioned information.

### Describe the solution you'd like.

I propose to introduce new reference documentation for Cython.

It can contain following information:
* data types supported by Cython (rework existing PR #https://github.com/cython/cython/pull/4980)
* tempita templeting system (it is already used by several project and currently is not documented)
* grammar definition
* Compiler options and Compiler directives from https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html
* Cython API (`Cython.Build.cythonize`, ...) reference documentation
* pure python mapping to Cython syntax (mainly coming from pure python tutorial)

### Describe alternatives you've considered.

To do nothing and use userguide for documenting reference information. My personal opinion is that this is suboptimal because it complicates reading for users not fully familiar with Cython.

### Additional context

See https://diataxis.fr/ for more rationale.

**Comment by matusvalo** [comment]
Please let me know about your thoughts. I can start working on it, I would start with list of data type which already has issue and PR...

**Comment by da-woods** [comment]
So at one point there was a reference section. The remnants of it still exist (https://cython.readthedocs.io/en/latest/src/reference/) but mostly just links where it's been moved to the user guide. I don't think it appears in the contents - I think the current version is just so old links to it still work.

I'm not sure why it was removed though...

**Comment by matusvalo** [comment]
> The remnants of it still exist

Uhm I honestly forget about it.

> I'm not sure why it was removed though...

It puzzles me too. I personally would like to build reference documentation but I don't want to go against previous decisions...

**Comment by da-woods** [comment]
I do like the idea of separating out this information into big comprehensive tables somewhere. One of the things I worry about with the existing documentation is that we try to over everything and people have to fight through a lot of text to get the basic idea.

The downside is that it's yet another place to search through (so I think it has to be linked well from the main documents). I think the user guide suffers a bit in places because some of the information is only in the tutorial, so this would add a third place.

This looks to have been the discussion where it was removed: https://github.com/cython/cython/issues/2245.

**Comment by matusvalo** [comment]
> The downside is that it's yet another place to search through

I think that this can be solved by simply adding links to the reference guide. E.g. userguide can talk how new type can be declared with simple example and have a simple link to reference guide e.g.

> The list of all supported data types is in type reference documentation(link)

**Comment by scoder** [comment]
The reason why the reference documentation was removed is that it was hugely redundant with the user guide and it was difficult to take decisions about what should be described where, so we ended up with incomplete documentation changes and outdated copies.

I don't think we need a completely separate reference documentation. We have the tutorials for introducing users to the concepts and the user guide for the completer scope and deeper details. What this ticket describes seems to fit mostly into the user guide. Tempita code generation is certainly a page on its own since it's an advanced topic. The Python syntax should be explained in combination with the .pyx file syntax, as we already do for many of the pages.
